src/Agent.Listener/Environment.cs

```diff
@@ -0,0 +1,338 @@
+                        await processInvoker.ExecuteAsync(
+                                    workingDirectory: string.Empty,
+                                    fileName: toolPath,
+                                    arguments: cap.CommandArgs,
+                                    environment: null,
+                                    cancellationToken: token);
+                        //TODO: should we check if toolOutput is a valid file path? (XPLAT does not check either)
+                        if (!string.IsNullOrEmpty(toolOutput))
+                        {
+                            capabilities.Add(cap.Name, toolOutput);
+                        }
+                    }
+                    catch (OperationCanceledException)
+                    {
+                        throw;
+                    }
+                    catch (Exception ex)
+                    {
+                        Trace.Error(ex);
+                    }
+                    finally
+                    {
+                        processInvoker.OutputDataReceived -= outputHandler;
+                    }
+                }
+
+                token.ThrowIfCancellationRequested();
+            }
+        }
+
+        private void GetFilteredEnvironmentVars(Dictionary<string, string> vars)
+        {
+            IDictionary envVars = System.Environment.GetEnvironmentVariables();
+
+            // Begin with ignoring env vars declared herein
+            var ignoredEnvVariables = new HashSet<string>(_ignoredEnvVariables);
+
+            // Also ignore env vars specified in the 'VSO_AGENT_IGNORE' env var
+            if (envVars.Contains(EnvIgnore))
+            {
+                var additionalIgnoredVars = ((string)envVars[EnvIgnore]).Split(',');
+                foreach (var ignored in additionalIgnoredVars)
+                {
+                    ignoredEnvVariables.Add(ignored.Trim());
+                }
+            }
+
+            // Get filtered env vars
+            foreach (DictionaryEntry envVar in envVars)
+            {
+                string varName = (string)envVar.Key;
+                string varValue = (string)envVar.Value ?? string.Empty;
+                if (!ignoredEnvVariables.Contains(varName) && varValue.Length < 1024)
+                {
+                    vars.Add(varName, varValue);
+                }
+            }
+        }
+
+        // Ensures existence of the environment file at the specified path, creating it if missing.
+        public void EnsureEnvFile()
+        {
+            string envFileName = IOUtil.GetEnvFilePath();
+            if (File.Exists(envFileName))
+            {
+                return;
+            }
+
+            var envVars = new Dictionary<string, string>();
+            GetFilteredEnvironmentVars(envVars);
+
+            var content = new StringBuilder(string.Empty);
+            foreach (var envvar in envVars)
+            {
+                content.Append($"{envvar.Key}={envvar.Value}");
```

Comment:
new lines and "=" in the values are not handled. If the value has multiple "=" it would be discarded. If it contains new lines, then it depends. I carefully crafted value may insert multiple environment variables. I don't know how common is to have new lines in environment value - XPLAT was not handling it either.